src/main/java/io/github/antishake/CircularBuffer.java

```diff
@@ -10,50 +16,66 @@
         -will maintain write pointer when updating with the latest values
  */
 public class CircularBuffer {
-  private int readPointer, writePointer = 0;
-  private int sizeBuffer;
-  private Coordinate[] circularbuffer;
-  private boolean bufferFullOnce = false;
-
-  public CircularBuffer(int size) {
-    sizeBuffer = size;
-    circularbuffer = new Coordinate[sizeBuffer];
-  }
-
-
-  public boolean isEmpty() {
-    return readPointer == writePointer;
-  }
-
-  public void add(Coordinate element) {
-    if (writePointer < sizeBuffer) {
-      if (!bufferFullOnce) {
-        circularbuffer[writePointer] = element;
-        writePointer++;
-      } else {
-        circularbuffer[readPointer] = element;
-        writePointer++;
-        readPointer++;
-      }
-    } else {
-      writePointer = 0;
-      circularbuffer[writePointer] = element;
-      writePointer++;
-      readPointer = readPointer + 2;
-      bufferFullOnce = true;
+  private int readPtr;
+  private int writePtr;
+  private ArrayList<Coordinate> buffer;
+  private int size;
+
+  public CircularBuffer(int maxSize) {
+    this.size = maxSize;
+    this.buffer = new ArrayList<>(maxSize);
+    for (int i = 0; i < maxSize; i++) {
+      this.buffer.add(null);
+    }
+  }
+
+  /**
+   * Add a coordinate to the buffer
+   */
+
+  public void add(Coordinate value) {
+    this.buffer.set(this.writePtr, value);
+    this.writePtr = ((this.writePtr + 1) % this.size);
+
+    if (this.readPtr == this.writePtr) {
+      this.readPtr = ((this.readPtr + 1) % this.size);
     }
   }
 
-  public Coordinate[] getElements() {
-    return circularbuffer;
+  public Coordinate read() {
+    if (this.readPtr == this.writePtr) {
+      return null;
+    }
+    Coordinate res = this.buffer.get(this.readPtr);
+    this.readPtr = ((this.readPtr + 1) % this.size);
+    return res;
+  }
+
+  public Coordinate[] readLatest(int len) {
+    throw new NotImplementedException();
+  }
+
+  public Coordinate[] readAll() {
+    ArrayList<Coordinate> res = new ArrayList<>();
+    Coordinate coord;
+    while ((coord = read()) != null) {
+      res.add(coord);
+    }
+    Coordinate[] arr = new Coordinate[res.size()];
+    return res.toArray(arr);
   }
 
   public int getReadPointer() {
-    return readPointer;
+    return readPtr;
   }
 
   public int getWritePointer() {
-    return writePointer;
+    return writePtr;
+  }
+
+  public Coordinate[] toArray() {
+    Coordinate[] array = new Coordinate[buffer.size()];
+    return buffer.toArray(array);
```

Comment:
If buffer is an ArrayList, we need not convert it to an array, as in AntiShake.java, the array is converted again to ArrayList.
As both the classes use ArrayList, we can remove the conversions.